File: Interviews/seatgeek/python/normalization/normalize.py
import argparse
from normalizer import Normalizer
import csv
import json

def tobool(s):
    """helper method to parse bools"""
    try:
        int_s = int(s)
        return bool(int_s)
    except:
        pass
    if s.lower()[0] == "t":
        return True
    elif s.lower()[0] == "f":
        return False
    else:
        raise ValueError("cannot find bool")

def toint(s):
    try:
        return int(s)
    except:
        if not s:
            return None
        raise ValueError("cannot convert to int")

def read_input(input_path):
    samples = []
    with open(input_path, 'rU') as f:
        reader = csv.DictReader(f)
        for row in reader:
            sample = {
                "input" : {"section" : row["section"], "row" : row["row"]},
                "expected" : {"section_id" : toint(row["n_section_id"]), "row_id" : toint(row["n_row_id"]), "valid" : tobool(row["valid"])}
            }
            samples.append(sample)
    return samples

def normalize_samples(normalizer, samples, verbose=False):
    matched = []
    for sample in samples:
        sid, rid, valid = normalizer.normalize(sample["input"]["section"], sample["input"]["row"])
        sample["output"] = {"section_id" : sid, "row_id" : rid, "valid" : valid}
        matched.append(sample)
    return matched

def output_samples(matched):
    """print this to the command line in json format
    the grading script will """
    for match in matched:
        print json.dumps(match)

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='grader for seatgeek SectionNormalization code test')
    parser.add_argument('--manifest', default=None, help='path to manifest file')
    parser.add_argument('--input', default=None, help='path to input file')
    parser.add_argument('--section', default=None, help='section input (for testing)')
    parser.add_argument('--row', default=None, help='row input (for testing)')

    args = parser.parse_args()

    assert args.manifest

    normalizer = Normalizer()
    normalizer.read_manifest(args.manifest)

    if args.section and args.row:
        section_id, row_id, valid = normalizer.normalize(args.section, args.row)
        print """
        Input:
            [section] {}\t[row] {}
        Output:
            [section_id] {}\t[row_id] {}
        Valid?:
            {}
        """.format(args.section, args.row, section_id, row_id, valid)

    elif args.input:
        samples = read_input(args.input)
        matched = normalize_samples(normalizer, samples, verbose=False)
        output_samples(matched)

File: Interviews/seatgeek/ruby/normalize.rb
require 'csv'
require 'json'
require 'optparse'
require './normalizer.rb'

# helper method to turn a string to a boolean
# understanding proper semantic value of the string
def to_boolean(s)
  s and !!s.match(/^(true|t|yes|y|1)$/i)
end

# helper method to turn a string to an int
# or nil (if not correctly formed)
def int_or_nil(s)
  Integer(s || '')
rescue ArgumentError
  nil
end

def read_input(input_path)
  samples = []
  CSV.foreach(input_path, headers: true) do |row|
    sample = {
      input: {section: row["section"], row: row["row"]},
      expected: {section_id: int_or_nil(row["n_section_id"]), row_id: int_or_nil(row["n_row_id"]), valid: to_boolean(row["valid"])}
    }
    samples << sample
  end
  samples
end

def normalize_samples(normalizer, samples, verbose=false)
  matched = samples.map do |sample|
    vals = normalizer.normalize(sample[:input][:section], sample[:input][:row])
    sid = vals[0]
    rid = vals[1]
    valid = vals[2]
    sample[:output] = {:section_id => int_or_nil(sid), :row_id => int_or_nil(rid), :valid => to_boolean(valid)}
    sample
  end
  matched
end

def output_samples(matched)
  matched.each do |match|
    puts JSON.dump(match)
  end
end

def main()
  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: normalize.rb [options]"

    opts.on('-mMANIFEST', '--manifest=MANIFEST', "path to manifest file") { |m| options[:manifest] = m }
    opts.on('-iINPUT', '--input=INPUT', "path to input file") { |i| options[:input] = i }
    opts.on('-sSECTION', '--section=SECTION', "section input (for testing)") { |s| options[:section] = s }
    opts.on('-rROW', '--row=ROW', "row input (for testing)") { |r| options[:row] = r }

  end.parse!
  puts options

  normalizer = Normalizer.new
  normalizer.read_manifest(options[:manifest])

  if options[:input]
    samples = read_input(options[:input])
    matched = normalize_samples(normalizer, samples, true)
    output_samples(matched)
  elsif options[:section] and options[:row]
    result = normalizer.normalize(options[:section], options[:row])
    s = "Input:\n    [section] #{options[:section]}\t[row] #{options[:row]}\nOutput:\n    [section_id] #{result[0]}\t[row_id] #{result[1]}\nValid?:\n    #{result[2]}"
    puts s
  end

end

# run main
main()
